flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java

```diff
@@ -140,15 +142,25 @@ public Catalog catalog() {
     return icebergCatalog;
   }
 
-  private Namespace toNamespace(String database) {
+  /** Append a new level to the base namespace */
+  private static Namespace appendLevel(Namespace baseNamespace, String newLevel) {
     String[] namespace = new String[baseNamespace.levels().length + 1];
     System.arraycopy(baseNamespace.levels(), 0, namespace, 0, baseNamespace.levels().length);
-    namespace[baseNamespace.levels().length] = database;
+    namespace[baseNamespace.levels().length] = newLevel;
     return Namespace.of(namespace);
   }
 
   TableIdentifier toIdentifier(ObjectPath path) {
-    return TableIdentifier.of(toNamespace(path.getDatabaseName()), path.getObjectName());
+    String objectName = path.getObjectName();
+    List<String> tableName = Splitter.on('$').splitToList(objectName);
+    if (tableName.size() == 2 && MetadataTableType.from(tableName.get(1)) != null) {
```

Comment:
`tableName.size() == 2 && MetadataTableType.from(tableName.get(1)) != null`
This condition seems to be sufficient to pick up the metadata table, and we don't need to add additional precondition check. We just need to make sure that the non-metadata table doesn't match the if branch, and it continues with the old logic. Or have I misunderstood what you're trying to say?

Reply:
after split, the list should have size of either 0 (regular table) or 2 (metadata table). `$` shouldn't be a valid char in other names.

if the list size is 1, 3, or any other numbers, it should be an illegal case.

Reply:
Thanks for your explanation, I have revised it. I think where we differ is whether the `$` can appear in the table name. In general, it is prohibited from being used in a table name. But technically, it can.
Of course, I think it's ok to ban it.
Thx.